src/flint/types/fmpz_mod.pyx

```diff
@@ -158,13 +184,123 @@ cdef class fmpz_mod(flint_scalar):
         res = fmpz_mod_is_one(self.val, self.ctx.val)
         return res == 1
 
+    def inverse(self, check=True):
+        r"""
+        Computes :math:`a^{-1} \pmod N`
+
+        When check=False, the solutions is assumed to exist and Flint will abort on
+        failure. 
+
+            >>> mod_ctx = fmpz_mod_ctx(163)
+            >>> mod_ctx(2).inverse()
+            fmpz_mod(82, 163)
+            >>> mod_ctx(2).inverse(check=False)
+            fmpz_mod(82, 163)
+        """
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
+
+        if check is False:
+            fmpz_mod_inv(res.val, self.val, self.ctx.val)
+            return res
+
+        cdef bint r
+        cdef fmpz one = fmpz.__new__(fmpz)
+        fmpz_one(one.val)
+
+        r = fmpz_mod_divides(
+            res.val, one.val, self.val, self.ctx.val
+        )
+        if r == 0:
+            raise ZeroDivisionError(f"{self} is not invertible modulo {self.ctx.modulus()}")
+
+        return res
+
+    def discrete_log(self, a, check=False):
+        """
+        Solve the discrete logarithm problem, using `self = g` as a base.
+        Assumes a solution, :math:`a = g^x \pmod p` exists.
+        
+        NOTE: Requires that the context modulus is prime.
+
+        TODO: This could instead be initalised as a class from a 
+        given base and the precomputations could be stored to allow 
+        faster computations for many discrete logs with the same base. 
```

Comment:
We already have a `.ctx` context object. Maybe this should be attached to the context object?

Reply:
It could be initialised at first use but stored on the context object after.

I don't know how much memory this uses compared to the size of the context object itself but if the base size of the struct is not large it could be stored inline as part of the ctx object or otherwise the ctx object could just have a pointer.

Reply:
We need to have the first discrete log on some base g, which is an fmpz_mod object, but certain parts of L only depend on the modulus and so yes, this could be on the context for sure.

Reply:
Oh, I see what you mean: there are two internal log calculations but if we stored the log of `g` we could reduce that to one.

So what you could add would be an object like
```python
log_g = fmpz_mod_log(g, ctx)
log_g(12)
```
I don't know how these things are used in practice and I don't have much sense for what the overhead would be of just always calling the internal log calculation twice (do these functions take a long time?).

Reply:
It's also doing way too much work, as it solves the discrete log for primative roots, but if I had some element of order 2, it would be much much faster to do Pohlig-Hellman in this subgroup.

TL;DR I think we shouldn't worry about optimising this in `python-flint` as it's not something state of the art in Flint itself. E.g. Sage calls a sub-exponential algorithm via Pari for Z/NZ, which is (partially) where all the time saving is here.